# Final Project - Exploratory Data Analysis of the Disney Datasets

## Submitted by Peng Zhang

## Introduction

### Question of Interest

This report investigates a question regarding the Disney datasets. The **question** is which **director** on the provided list has produced the Disney movies with the highest average inflation-adjusted gross revenue. This is interesting because people would like to know which director has the most successful box office history on an average basis. The expected answer is **Wolfgang Reitherman** since his name appears mostly on the provided list; however, the below analysis will discover whether the assumption is correct or not.

### Dataset Description

The Disney dataset is composed of 5 tables, `disney-characters.csv`, `disney-director.csv`, `disney-voice-actors.csv`, `disney_revenue_1991-2016.csv`, and `disney_movies_total_gross.csv`. Each table is stored in a `.csv file` and contains different information about characters, directors, voice actors, Disney revenues from 1991 to 2016, and gross revenues of Disney movies from 1937 to 2016. In order to answer the above question, two tables, `disney-director.csv` and `disney_movies_total_gross.csv` will be utilized for the analysis.

* **disney-director.csv**

    - This file contains information on the Disney dataset, including the name of a movie and the director's name.

* **disney_movies_total_gross.csv**

    - This file contains information on the Disney dataset, including the title (name) of a movie, the release date, the movie genre, the MPAA rating, the total gross revenue generated from the movie, and the inflation-adjusted gross revenue that removes the effects of inflation.

## Methods and Results

To start the analysis, the initial step is to import all the required libraries, load the two tables needed from the data folder and save them as two objects, **disney_director** and **disney_movies_total_gross**, respectively.

In [1]:
# Import libraries required for this project
import pandas as pd
import altair as alt

In [2]:
# Import all the required data files
disney_director = pd.read_csv('data/disney-director.csv')
disney_movies_total_gross = pd.read_csv('data/disney_movies_total_gross.csv')

Output the first 5 rows of the two dataframes as below to get a general idea how the tables look like.

In [3]:
disney_director.head()

name        director
0  Snow White and the Seven Dwarfs      David Hand
1                        Pinocchio  Ben Sharpsteen
2                         Fantasia    full credits
3                            Dumbo  Ben Sharpsteen
4                            Bambi      David Hand

In [4]:
disney_movies_total_gross.head()

movie_title  release_date      genre MPAA_rating  \
0  Snow White and the Seven Dwarfs  Dec 21, 1937    Musical           G   
1                        Pinocchio   Feb 9, 1940  Adventure           G   
2                         Fantasia  Nov 13, 1940    Musical           G   
3                Song of the South  Nov 12, 1946  Adventure           G   
4                       Cinderella  Feb 15, 1950      Drama           G   

    total_gross inflation_adjusted_gross  
0  $184,925,485           $5,228,953,251  
1   $84,300,000           $2,188,229,052  
2   $83,320,000           $2,187,090,808  
3   $65,000,000           $1,078,510,579  
4   $85,000,000             $920,608,730

Extract some informatin about the **disney_director** dataframe and **disney_movies_total_gross** dataframe.

In [5]:
disney_director.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56 entries, 0 to 55
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   name      56 non-null     object
 1   director  56 non-null     object
dtypes: object(2)
memory usage: 1.0+ KB


As indicated as above, the **disney_director** dataframe has 56 rows and 2 columns to list the director's name for 56 Disney movies.

In [6]:
disney_movies_total_gross.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 579 entries, 0 to 578
Data columns (total 6 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   movie_title               579 non-null    object
 1   release_date              579 non-null    object
 2   genre                     562 non-null    object
 3   MPAA_rating               523 non-null    object
 4   total_gross               579 non-null    object
 5   inflation_adjusted_gross  579 non-null    object
dtypes: object(6)
memory usage: 27.3+ KB


As shown above, the **disney_movies_total_gross** dataframe has 579 rows and 6 columns. The table has disclosed the release date, the total gross revenue, and the inflation-adjusted gross revenue for each of the listed Disney movies. In addition, most of the movies have information such as the film genre and the MPAA rating. There are 17 out of the 579 movies without identifying the film genre and 56 movies are missing the MPAA rating.

Prior to focusing on the defined question, some people may be curious about Disney movies' revenue-generating capabilities over the years. It would be a meaningful approach to take a glance through the **disney_movies_total_gross** dataframe and calculate the total annual gross revenue from the listed Disney movies. For comparison purposes, the inflation-adjusted value will be considered rather than the gross revenue for the particular year.

Here a modification will be performed firstly to transform the column of "release_date" so that it adopts a datetime64 dtype and then extract the year from the datetime value.

In [7]:
# Convert the column of "release_date" to a datetime64 dtype
date_modified = disney_movies_total_gross.assign(release_date = pd.to_datetime(disney_movies_total_gross['release_date']))

# Add a new column to extract the year from the datetime value 
date_modified = date_modified.assign(year = date_modified['release_date'].dt.year)
date_modified

movie_title release_date      genre MPAA_rating  \
0    Snow White and the Seven Dwarfs   1937-12-21    Musical           G   
1                          Pinocchio   1940-02-09  Adventure           G   
2                           Fantasia   1940-11-13    Musical           G   
3                  Song of the South   1946-11-12  Adventure           G   
4                         Cinderella   1950-02-15      Drama           G   
..                               ...          ...        ...         ...   
574         The Light Between Oceans   2016-09-02      Drama       PG-13   
575                   Queen of Katwe   2016-09-23      Drama          PG   
576                   Doctor Strange   2016-11-04  Adventure       PG-13   
577                            Moana   2016-11-23  Adventure          PG   
578     Rogue One: A Star Wars Story   2016-12-16  Adventure       PG-13   

      total_gross inflation_adjusted_gross  year  
0    $184,925,485           $5,228,953,251  1937  
1     $84,300,000           $2,188,229,052  1940  
2     $83,320,000           $2,187,090,808  1940  
3     $65,000,000           $1,078,510,579  1946  
4     $85,000,000             $920,608,730  1950  
..            ...                      ...   ...  
574   $12,545,979              $12,545,979  2016  
575    $8,874,389               $8,874,389  2016  
576  $232,532,923             $232,532,923  2016  
577  $246,082,029             $246,082,029  2016  
578  $529,483,936             $529,483,936  2016  

[579 rows x 7 columns]

Additional modification is required to clean up the currency data from the column of "inflation_adjusted_gross" and convert them into an appropriate format for numeric operations. A python script with a custom function, <font color='blue'>clean_currency</font>, has been created for this project. This function is a simple tool to remove currency symbol and delimiters from dollar values.

The modification for the column of "inflation_adjusted_gross" will be executed through the following codes:

In [9]:
# Import the custom function from the Python script
from script import clean_currency

# Apply the custom function, clean_currency, to remove "$" and "," then convert the dollar value data to the type of float
gross_modified = date_modified.assign(inflation_adjusted_gross = date_modified['inflation_adjusted_gross'].apply(clean_currency).astype('float'))
gross_modified

movie_title release_date      genre MPAA_rating  \
0    Snow White and the Seven Dwarfs   1937-12-21    Musical           G   
1                          Pinocchio   1940-02-09  Adventure           G   
2                           Fantasia   1940-11-13    Musical           G   
3                  Song of the South   1946-11-12  Adventure           G   
4                         Cinderella   1950-02-15      Drama           G   
..                               ...          ...        ...         ...   
574         The Light Between Oceans   2016-09-02      Drama       PG-13   
575                   Queen of Katwe   2016-09-23      Drama          PG   
576                   Doctor Strange   2016-11-04  Adventure       PG-13   
577                            Moana   2016-11-23  Adventure          PG   
578     Rogue One: A Star Wars Story   2016-12-16  Adventure       PG-13   

      total_gross  inflation_adjusted_gross  year  
0    $184,925,485              5.228953e+09  1937  
1     $84,300,000              2.188229e+09  1940  
2     $83,320,000              2.187091e+09  1940  
3     $65,000,000              1.078511e+09  1946  
4     $85,000,000              9.206087e+08  1950  
..            ...                       ...   ...  
574   $12,545,979              1.254598e+07  2016  
575    $8,874,389              8.874389e+06  2016  
576  $232,532,923              2.325329e+08  2016  
577  $246,082,029              2.460820e+08  2016  
578  $529,483,936              5.294839e+08  2016  

[579 rows x 7 columns]

Then move forward with grouping the modified dateframe by year and then compute the total annual gross revenue from the listed Disney movies:

In [10]:
# Group by year and calculate annual gross from the movies
gross_modified_group = pd.DataFrame(gross_modified.groupby('year')['inflation_adjusted_gross'].sum())

# Reset the index so it can plot using altair
gross_modified_group = gross_modified_group.reset_index()
gross_modified_group

year  inflation_adjusted_gross
0   1937              5.228953e+09
1   1940              4.375320e+09
2   1946              1.078511e+09
3   1950              9.206087e+08
4   1954              5.282800e+08
5   1955              1.236036e+09
6   1959              2.150583e+07
7   1961              1.797807e+09
8   1962              1.095816e+08
9   1963              1.538708e+08
10  1967              7.896123e+08
11  1968              1.386127e+08
12  1970              3.563622e+08
13  1971              9.130545e+07
14  1975              1.312469e+08
15  1977              3.636592e+08
16  1979              1.203774e+08
17  1980              4.356021e+07
18  1981              1.816369e+08
19  1982              1.107312e+08
20  1983              1.005495e+08
21  1984              1.781327e+08
22  1985              1.731242e+08
23  1986              5.774724e+08
24  1987              1.186650e+09
25  1988              9.868419e+08
26  1989              1.097851e+09
27  1990              1.226461e+09
28  1991              1.221190e+09
29  1992              1.736350e+09
30  1993              1.427313e+09
31  1994              2.140691e+09
32  1995              2.188599e+09
33  1996              2.157239e+09
34  1997              1.474848e+09
35  1998              2.189031e+09
36  1999              1.981583e+09
37  2000              1.711676e+09
38  2001              1.296292e+09
39  2002              1.680600e+09
40  2003              2.171245e+09
41  2004              1.570219e+09
42  2005              1.273983e+09
43  2006              1.834745e+09
44  2007              1.757480e+09
45  2008              1.135356e+09
46  2009              1.359909e+09
47  2010              1.622535e+09
48  2011              1.172579e+09
49  2012              1.537669e+09
50  2013              1.887864e+09
51  2014              1.560791e+09
52  2015              2.495663e+09
53  2016              2.873393e+09

Using the above table, a bar plot will be generated to visualize the result.

In [11]:
# Use altair to generate a bar plot
annual_gross_plot = (
    alt.Chart(gross_modified_group, width=500, height=300)
    .mark_bar()
    .encode(
        x=alt.X("year:O", title="Year"),
        y=alt.Y("inflation_adjusted_gross:Q", title="Annual Inflation-adjusted Gross Revenue ($)"),
    )
    .properties(title="Figure 1 Disney Movies Annual Inflation-adjusted Gross Revenue by Year")
)
annual_gross_plot

alt.Chart(...)

Based on the above plot, Disney movies had an extremely successful history back to the 1930s and 1940s. That was contributed from a few highest-grossing sound films including "Snow White and the Seven Dwarfs" at the time. Between the 1960s and 1985 Disney movies had experienced a long period with much lower financial performances. And after 1985 till 2016, the annual inflation-adjusted gross revenues appear to be a relatively stable trend over the 30 years although the data fluctuated within a certain range.

Now it is time to back to the original **question**. Which director on the provided list has produced the Disney movies with the highest average inflation-adjusted gross revenue? The above created **gross_modified** dataframe will be merged with the **disney_director** dataframe. Name the new dataframe **gross_by_director**. Then to compute the average inflation-adjusted gross revenue for each director.

Since the dataframe **disney_director** only shows 56 movies with their director information while another dataframe **gross_modified** has 579 rows of movies, it suspects that some movies included in the dataframe **gorss_modified** may not have available director information. Therefore, the "left" argument will be used as the type of join within the **merge** function to verify if some movies have missing data for the director's name. 

In [12]:
# Merge the two dataframes and only output the rows that are in the disney_director dataframe
gross_by_director = disney_director.merge(gross_modified, left_on='name', right_on='movie_title', how='left')
print (gross_by_director.shape)
gross_by_director

(60, 9)


name             director  \
0          Snow White and the Seven Dwarfs           David Hand   
1                                Pinocchio       Ben Sharpsteen   
2                                 Fantasia         full credits   
3                                    Dumbo       Ben Sharpsteen   
4                                    Bambi           David Hand   
5                           Saludos Amigos          Jack Kinney   
6                     The Three Caballeros      Norman Ferguson   
7                          Make Mine Music          Jack Kinney   
8                       Fun and Fancy Free          Jack Kinney   
9                              Melody Time       Clyde Geronimi   
10  The Adventures of Ichabod and Mr. Toad          Jack Kinney   
11                              Cinderella      Wilfred Jackson   
12                              Cinderella      Wilfred Jackson   
13                     Alice in Wonderland       Clyde Geronimi   
14                               Peter Pan       Hamilton Luske   
15                      Lady and the Tramp       Hamilton Luske   
16                         Sleeping Beauty       Clyde Geronimi   
17                          101 Dalmatians  Wolfgang Reitherman   
18                          101 Dalmatians  Wolfgang Reitherman   
19                  The Sword in the Stone  Wolfgang Reitherman   
20                         The Jungle Book  Wolfgang Reitherman   
21                         The Jungle Book  Wolfgang Reitherman   
22                         The Jungle Book  Wolfgang Reitherman   
23                          The Aristocats  Wolfgang Reitherman   
24                              Robin Hood  Wolfgang Reitherman   
25  The Many Adventures of Winnie the Pooh  Wolfgang Reitherman   
26                            The Rescuers  Wolfgang Reitherman   
27                   The Fox and the Hound          Art Stevens   
28                      The Black Cauldron           Ted Berman   
29               The Great Mouse Detective         Ron Clements   
30                        Oliver & Company      George Scribner   
31                      The Little Mermaid         Ron Clements   
32                 The Rescuers Down Under         Mike Gabriel   
33                    Beauty and the Beast       Gary Trousdale   
34                                 Aladdin         Ron Clements   
35                           The Lion King         Roger Allers   
36                              Pocahontas         Mike Gabriel   
37             The Hunchback of Notre Dame       Gary Trousdale   
38                                Hercules         Ron Clements   
39                                   Mulan           Barry Cook   
40                                  Tarzan           Chris Buck   
41                           Fantasia 2000         full credits   
42                                Dinosaur         Ralph Zondag   
43                The Emperor's New Groove          Mark Dindal   
44               Atlantis: The Lost Empire       Gary Trousdale   
45                           Lilo & Stitch        Chris Sanders   
46                         Treasure Planet         Ron Clements   
47                            Brother Bear        Robert Walker   
48                       Home on the Range            Will Finn   
49                          Chicken Little          Mark Dindal   
50                      Meet the Robinsons  Stephen J. Anderson   
51                                    Bolt       Chris Williams   
52               The Princess and the Frog         Ron Clements   
53                                 Tangled         Nathan Greno   
54                         Winnie the Pooh  Stephen J. Anderson   
55                          Wreck-It Ralph           Rich Moore   
56                                  Frozen           Chris Buck   
57                              Big Hero 6             Don Hall   
58                                Zootopia         Byron Howard   
59                

From the above output, 11 out of 60 movies are missing from the **gorss_modified** dataframe. **The reason for missing data from the provided Disney Datasets will not be examined in this report.** To disregard the NaN values and continue the analysis, the new dataframe **gross_by_director** will be re-generated using the "inner" argument within the merge function.

In [13]:
# Merge the two dataframes and only keep the rows with identifying column values that are present in both dataframes
gross_by_director = disney_director.merge(gross_modified, left_on='name', right_on='movie_title', how='inner')
print (gross_by_director.shape)
gross_by_director

(49, 9)


name             director  \
0          Snow White and the Seven Dwarfs           David Hand   
1                                Pinocchio       Ben Sharpsteen   
2                                 Fantasia         full credits   
3                               Cinderella      Wilfred Jackson   
4                               Cinderella      Wilfred Jackson   
5                      Alice in Wonderland       Clyde Geronimi   
6                       Lady and the Tramp       Hamilton Luske   
7                          Sleeping Beauty       Clyde Geronimi   
8                           101 Dalmatians  Wolfgang Reitherman   
9                           101 Dalmatians  Wolfgang Reitherman   
10                  The Sword in the Stone  Wolfgang Reitherman   
11                         The Jungle Book  Wolfgang Reitherman   
12                         The Jungle Book  Wolfgang Reitherman   
13                         The Jungle Book  Wolfgang Reitherman   
14                          The Aristocats  Wolfgang Reitherman   
15  The Many Adventures of Winnie the Pooh  Wolfgang Reitherman   
16                            The Rescuers  Wolfgang Reitherman   
17                   The Fox and the Hound          Art Stevens   
18                      The Black Cauldron           Ted Berman   
19               The Great Mouse Detective         Ron Clements   
20                        Oliver & Company      George Scribner   
21                      The Little Mermaid         Ron Clements   
22                 The Rescuers Down Under         Mike Gabriel   
23                    Beauty and the Beast       Gary Trousdale   
24                                 Aladdin         Ron Clements   
25                           The Lion King         Roger Allers   
26                              Pocahontas         Mike Gabriel   
27             The Hunchback of Notre Dame       Gary Trousdale   
28                                Hercules         Ron Clements   
29                                   Mulan           Barry Cook   
30                                  Tarzan           Chris Buck   
31                                Dinosaur         Ralph Zondag   
32                The Emperor's New Groove          Mark Dindal   
33               Atlantis: The Lost Empire       Gary Trousdale   
34                           Lilo & Stitch        Chris Sanders   
35                         Treasure Planet         Ron Clements   
36                            Brother Bear        Robert Walker   
37                       Home on the Range            Will Finn   
38                          Chicken Little          Mark Dindal   
39                      Meet the Robinsons  Stephen J. Anderson   
40                                    Bolt       Chris Williams   
41               The Princess and the Frog         Ron Clements   
42                                 Tangled         Nathan Greno   
43                         Winnie the Pooh  Stephen J. Anderson   
44                          Wreck-It Ralph           Rich Moore   
45                                  Frozen           Chris Buck   
46                              Big Hero 6             Don Hall   
47                                Zootopia         Byron Howard   
48                                   Moana         Ron Clements   

                               movie_title release_date      genre  \
0          Snow White and the Seven Dwarfs   1937-12-21    Musical   
1                                Pinocchio   1940-02-09  Adventure   
2                                 Fantasia   1940-11-13    Musical   
3                               Cinderella   1950-02-15      Drama   
4                               Cinderella   2015-03-13      Drama   
5                      Alice in Wonderland   2010-03-05  Adventure   
6                       Lady and the Tramp   1955-06-22      Drama   
7                          Sleeping Beauty   1959-01-29      Drama   
8                           101 Dalmatians   1961-01-25  

To answer the question, the next is to group by the directors and then compute the average value of inflation-adjusted gross revenue from each director's movies.

In [14]:
# Group by directors and compute the average value of inflation-adjusted gross revenue
gross_by_director_mean = pd.DataFrame(gross_by_director.groupby('director')['inflation_adjusted_gross'].mean().sort_values(ascending=False))

# Reset the index for using altair
gross_by_director_mean = gross_by_director_mean.reset_index()
gross_by_director_mean

director  inflation_adjusted_gross
0            David Hand              5.228953e+09
1        Ben Sharpsteen              2.188229e+09
2          full credits              2.187091e+09
3        Hamilton Luske              1.236036e+09
4          Roger Allers              7.616409e+08
5       Wilfred Jackson              5.608800e+08
6   Wolfgang Reitherman              3.814355e+08
7            Chris Buck              3.494487e+08
8          Byron Howard              3.412682e+08
9              Don Hall              2.292492e+08
10       Gary Trousdale              2.263982e+08
11           Barry Cook              2.168078e+08
12         Ralph Zondag              2.154390e+08
13         Nathan Greno              2.143885e+08
14        Chris Sanders              2.115067e+08
15           Rich Moore              2.003550e+08
16       Clyde Geronimi              1.892847e+08
17         Ron Clements              1.884214e+08
18         Mike Gabriel              1.650838e+08
19          Mark Dindal              1.573720e+08
20       Chris Williams              1.337025e+08
21          Art Stevens              1.331189e+08
22        Robert Walker              1.192183e+08
23      George Scribner              1.022545e+08
24  Stephen J. Anderson              7.411823e+07
25            Will Finn              6.791017e+07
26           Ted Berman              5.055314e+07

It seems that in fact **David Hand** has produced the Disney movie with the highest average inflation-adjusted gross revenue although he only has one Disney movie listed on the provided dataset. This is different from the initial assumption of it being **Wolfgang Reitherman**.

The last step is to visualize the above finding by ploting a bar chart.

In [15]:
# Use altair to generate a bar plot
gross_by_director_mean_plot = (
    alt.Chart(gross_by_director_mean, width=500, height=300)
    .mark_bar()
    .encode(
        x=alt.X("director:N", sort="-y", title="Director"),
        y=alt.Y("inflation_adjusted_gross:Q", title="Average Inflation-adjusted Gross Revenue ($)"),
    )
    .properties(title="Figure 2 Average Inflation-adjusted Gross Revenue from Director's Disney Movies")
)
gross_by_director_mean_plot

alt.Chart(...)

## Discussions

In this analysis, two tables from the Disney datasets, `disney-director.csv` and `disney_movies_total_gross.csv`, have been utilized to discover which director on the provided list has produced the Disney movies with the highest average inflation-adjusted gross revenue. Before answering this question, a quick exploratory data analysis has been completed to review Disney movies' revenue generating capabilities over the years. From the first visualization, it indicates that Disney movies had a significate success back to the 1930s and 1940s, and then kept quite lower financial perforamce between the 1960s and 1985. After 1985 till 2016, the annual gross revenues shows a relatively stable trend over the 30 years.

The expected director with the highest average inflation-adjusted gross revenue was **Wolfgang Reitherman**; however, the data computing result reveals that **David Hand** actually holds the record. Although Mr. Hand only has two Disney movies included in the `disney-director.csv` table, his "Snow White and the Seven Dwarfs" generated over 5 billion US Dollars gross revenue with the consideration of inflation adjustment.

It should be noted that the above result is based on the available data from the provided Disney Datasets. As mentioned in the report, some information seems missing in the original datasets that may imply a **limitation** of the result. Deep analysis or access to more data sources could be considered if people would like to conduct further investigation.

Some people may argue that the defined question and answer only advise Disney directors' overall financial performances and popularity. The rank cannot precisely reflect directors' excellences in their particular field because some factors, such as the film genre and MPAA_rating, have not been included in the considerations. Public tastes vary from different categories of movies. A brilliant animated movie director could have less economic successes if they are focusing on a particular film genre that does not have the largest audience pool. So the future analysis can dig more exploratory questions to find out the best directors within diverse categories of Disney movies.

## References

### Resources Used

Not all the work in this project is original. The following resources have been used as references.

* <font color='blue'>Data Source</font>

The Disney datesets provided by UBC Programming in Python for Data Science course.

* <font color='blue'>Sample Report</font>

The sample report, **Exploratory data analysis of the Lego datasets** provided by UBC Programming in Python for Data Science course.

* <font color='blue'>Report Writing</font>

Inspiration for defining questions and writing data analysis report from the publications, *Art of Data Science*, by Roger D. Peng and Elizabeth Matsui, and *Report Writing for Data Science in R*, by Roger D. Peng.

Knowledge for formating text from the web article, **Lesson 2. Format Text In Jupyter Notebook With Markdown** by Jenny Palomino and Leah Wasser, retrieved on August 28, 2021 from <a href="https://www.earthdatascience.org/courses/intro-to-earth-data-science/file-formats/use-text-files/format-text-with-markdown-jupyter-notebook/" target="_blank">Earth Data Analytics Online Certificate</a>.

* <font color='blue'>Python Techniques</font>

Learned from UBC Programming in Python for Data Science course and self-learning resources from the internet.

<h2><center>**End of the Report**</center></h2>